Userland/Libraries/LibLocate/LocateDb.cpp

```diff
@@ -0,0 +1,221 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibLocate/LocateDb.h>
+
+namespace Locate {
+
+LocateDb::LocateDb(String path, LocateDbMode mode)
+{
+    if (!Core::File::exists(LOCATE_DB_PATH)) {
+        VERIFY(Core::File::ensure_parent_directories(LOCATE_DB_PATH));
+        auto file_handle = Core::File::open(LOCATE_DB_PATH, Core::IODevice::WriteOnly, 0640);
+        if (file_handle.is_error()) {
+            outln("Got error while trying to create locate db file: {}", file_handle.error());
+            VERIFY_NOT_REACHED();
+        }
+        file_handle.value()->close();
+        chown(LOCATE_DB_PATH, 0, 18);
+    }
+
+    m_file_handle = fopen(path.characters(), (mode == LocateDbMode::write ? "wb" : "rb"));
+    fseek(m_file_handle, 0, SEEK_END);
+    m_file_size = ftell(m_file_handle);
+    fseek(m_file_handle, 0, SEEK_SET);
+}
+
+LocateDb::~LocateDb()
+{
+    fclose(m_file_handle);
+}
+
+void LocateDb::write_header()
+{
+    // Format: magic string + version.
+    String db_header = String::formatted("{}|{}", "locatedb", 1);
+    fwrite(db_header.characters(), sizeof(uint8_t), db_header.length() + 1, m_file_handle);
+}
+
+void LocateDb::write_directory(DirectoryInfo& dir_info)
+{
+    struct DirStartEndHeader dir_header {
+        .type = (uint8_t)ChunkType::dir_start,
+        .db_id = dir_info.db_id,
+        .parent_db_id = dir_info.parent_db_id,
+        .path_size = dir_info.path.length() + 1
+    };
+    fwrite((char*)&dir_header, sizeof(struct DirStartEndHeader), 1, m_file_handle);
+    fwrite(dir_info.path.characters(), sizeof(uint8_t), dir_info.path.length() + 1, m_file_handle);
+
+    for (auto& child : dir_info.children) {
+        struct ChildFileData cur_child {
+            .type = (uint8_t)(child->type == FileType::file ? ChunkType::file : ChunkType::dir),
+            .db_id = child->db_id,
+            .parent_db_id = child->parent_db_id,
+            .user_owner = child->user_owner,
+            .group_owner = child->group_owner,
+            .st_mode = child->st_mode,
+            .name_size = child->name.length() + 1
+        };
+
+        fwrite((char*)&cur_child, sizeof(struct ChildFileData), 1, m_file_handle);
+        fwrite(child->name.characters(), sizeof(uint8_t), child->name.length() + 1, m_file_handle);
+    }
+
+    dir_header.type = (uint8_t)ChunkType::dir_end;
+    fwrite((char*)&dir_header, sizeof(struct DirStartEndHeader), 1, m_file_handle);
+    fwrite(dir_info.path.characters(), sizeof(uint8_t), dir_info.path.length() + 1, m_file_handle);
+}
+
+bool LocateDb::verify_header()
+{
+    char header[] = "locatedb|1";
+    char buffer[strlen(header) + 1];
+    size_t bytes_read = fread(&buffer, sizeof(char), strlen("locatedb|1") + 1, m_file_handle);
+    VERIFY(bytes_read > 0);
```

Comment:
what if it `== 0`?
why not just return `false`?

Reply:
Might have been a leftover from debugging; although, the CI won't like freads where the read_bytes value isn't read (since gcc will generate warnings for functions where a value is being returned but that value isn't stored in a var, which in turn has to be used in some way to not generate a warning). Anyway, moving that check into the "if"-check instead.